### PR TITLE
Moving class level methods to instance methods for Repositories

### PR DESCRIPTION
**Pull request by pascalbetz**
See https://github.com/hanami/model/issues/290

* Keep class level accessors for _adapter_ and _collection_
* Add instance level accessors for _adapter_ and _collection_ that use
  class level accessors
* Move class to instance level methods
* Change all tests and sample code to reflect changes

### CODE REVIEW

**Comment by davydovanton** [comment]
good job :+1:

**Comment by hlegius** [comment]
:ok_hand:

**Comment by joneslee85** [comment]
:+1:

**Comment by joneslee85** [comment]
@pascalbetz please rebase

**Comment by pascalbetz** [comment]
@joneslee85 rebase done.

File: lib/hanami/model.rb [hunk 1]
```diff
@@ -11,6 +11,78 @@ module Hanami
...
   # @since 0.1.0
   module Model
+    # Error for non persisted entity
+    # It's raised when we try to update or delete a non persisted entity.
+    #
+    # @since 0.1.0
+    #
+    # @see Hanami::Repository.new.update
+    class NonPersistedEntityError < Hanami::Model::Error
+    end
```
Comment (joneslee85) [inline lib/hanami/model.rb]: I think this is not related to this PR. Seems a bad rebase?

Comment (hlegius) [inline lib/hanami/model.rb]: Yep. It was moved to `model/error.rb`. You can safely remove these lines.

File: lib/hanami/repository.rb [hunk 2]
```diff
@@ -208,16 +209,33 @@ module ClassMethods
       #   end
       #
       #   UserRepository.adapter = adapter
+      # @since 0.5.0
+      # @api private
       def adapter=(adapter)
         @adapter = adapter
       end
 
-      # @since 0.5.0
```
Comment (joneslee85) [inline lib/hanami/repository.rb]: Shouldn't the comments removed?

File: lib/hanami/repository.rb [hunk 3]
```diff
@@ -208,16 +209,33 @@ module ClassMethods
...
       def adapter
         @adapter
       end
 
+    end
+
+    module InstanceMethods
+      # Get the currently configured adapter
+      #
+      # Returns the adapter that is specified on the class level
```
Comment (joneslee85) [inline lib/hanami/repository.rb]: I do not think we need this comment line since the code reveals the intent well enough

File: lib/hanami/repository.rb (hunk 2 above)
Comment (pascalbetz) [inline lib/hanami/repository.rb]: Good question. Actually I was thinking about removing/resetting all the ```@version``` comments as well for the ```InstanceMethods``` methods that were previous ```ClassMethods```

About removing the comments: it's still valid. Just not the preferred way anymore. But until we have decided what the new preferred way is, i'd leave the comments.

File: test/integration/repository_test.rb [hunk 4]
```diff
@@ -589,7 +611,7 @@ def updated_at
       end
 
       it "raises an exception because memory adapter doesn't support fetch" do
-        -> { ArticleRepository.find_raw }.must_raise NotImplementedError
+        -> { ArticleRepository.new.find_raw }.must_raise NotImplementedError
```
Comment (joneslee85) [inline test/integration/repository_test.rb]: can we asset the exception message too?

**Comment by joneslee85** [comment]
except comments above, this is a good PR

File: lib/hanami/repository.rb (hunk 2 above)
Comment (joneslee85) [inline lib/hanami/repository.rb]: I see. Fair point

**Comment by joneslee85** [comment]
cc @AlfonsoUceda

File: lib/hanami/repository.rb [hunk 5]
```diff
@@ -208,16 +209,29 @@ module ClassMethods
...
-      # @since 0.5.0
-      # @api private
       def adapter
         @adapter
       end
 
+    end
+
+    module InstanceMethods
+      # Get the currently configured adapter
```
Comment (AlfonsoUceda) [inline lib/hanami/repository.rb]: could you add `@since x.x.x` and below?

**Comment by AlfonsoUceda** [comment]
:+1: we will see where we end with this changes :D

File: lib/hanami/repository.rb (hunk 2 above)
Comment (jodosha) [inline lib/hanami/repository.rb]: @pascalbetz @joneslee85 Yes, class => instance methods should be updated with `@version x.x.x`.

**Comment by jodosha** [comment]
@pascalbetz This looks good to me. Merging :+1:
